File: python/syntax.py
"""While 言語の抽象構文木 (AST)。

コンパイラはこの木を入力として仮想スタック命令へ翻訳する。
型は OCaml 版 (syntax.ml) と同じ構成にしてある。
"""

from __future__ import annotations

from dataclasses import dataclass


# --- 算術式 -----------------------------------------------------------------

@dataclass(frozen=True)
class Var:
    name: str


@dataclass(frozen=True)
class Num:
    n: int


@dataclass(frozen=True)
class Add:
    lhs: Arith
    rhs: Arith


@dataclass(frozen=True)
class Sub:
    lhs: Arith
    rhs: Arith


@dataclass(frozen=True)
class Mul:
    lhs: Arith
    rhs: Arith


@dataclass(frozen=True)
class Div:
    lhs: Arith
    rhs: Arith


Arith = Var | Num | Add | Sub | Mul | Div


# --- 条件式 -----------------------------------------------------------------

@dataclass(frozen=True)
class TrueConst:
    pass


@dataclass(frozen=True)
class FalseConst:
    pass


@dataclass(frozen=True)
class Not:
    pred: Pred


@dataclass(frozen=True)
class And:
    lhs: Pred
    rhs: Pred


@dataclass(frozen=True)
class Or:
    lhs: Pred
    rhs: Pred


@dataclass(frozen=True)
class LT:
    lhs: Arith
    rhs: Arith


@dataclass(frozen=True)
class LE:
    lhs: Arith
    rhs: Arith


@dataclass(frozen=True)
class GT:
    lhs: Arith
    rhs: Arith


@dataclass(frozen=True)
class GE:
    lhs: Arith
    rhs: Arith


@dataclass(frozen=True)
class EQ:
    lhs: Arith
    rhs: Arith


Pred = TrueConst | FalseConst | Not | And | Or | LT | LE | GT | GE | EQ


# --- 文 ---------------------------------------------------------------------

@dataclass(frozen=True)
class Assign:
    name: str
    expr: Arith


@dataclass(frozen=True)
class Skip:
    pass


@dataclass(frozen=True)
class Block:
    body: Stmt


@dataclass(frozen=True)
class Seq:
    first: Stmt
    second: Stmt


@dataclass(frozen=True)
class While:
    pred: Pred
    body: Stmt


@dataclass(frozen=True)
class Print:
    expr: Arith | Pred


@dataclass(frozen=True)
class If:
    pred: Pred
    then_branch: Stmt
    else_branch: Stmt


Stmt = Assign | Skip | Block | Seq | While | Print | If


def string_of_arith(a: Arith) -> str:
    match a:
        case Var(name):
            return f"Var({name})"
        case Num(n):
            return f"Num({n})"
        case Add(lhs, rhs):
            return f"Add({string_of_arith(lhs)}, {string_of_arith(rhs)})"
        case Sub(lhs, rhs):
            return f"Sub({string_of_arith(lhs)}, {string_of_arith(rhs)})"
        case Mul(lhs, rhs):
            return f"Mul({string_of_arith(lhs)}, {string_of_arith(rhs)})"
        case Div(lhs, rhs):
            return f"Div({string_of_arith(lhs)}, {string_of_arith(rhs)})"


def string_of_pred(p: Pred) -> str:
    match p:
        case TrueConst():
            return "true"
        case FalseConst():
            return "false"
        case Not(pred):
            return f"not {string_of_pred(pred)}"
        case And(lhs, rhs):
            return f"({string_of_pred(lhs)} and {string_of_pred(rhs)})"
        case Or(lhs, rhs):
            return f"({string_of_pred(lhs)} or {string_of_pred(rhs)})"
        case LT(lhs, rhs):
            return f"{string_of_arith(lhs)} < {string_of_arith(rhs)}"
        case LE(lhs, rhs):
            return f"{string_of_arith(lhs)} <= {string_of_arith(rhs)}"
        case GT(lhs, rhs):
            return f"{string_of_arith(lhs)} > {string_of_arith(rhs)}"
        case GE(lhs, rhs):
            return f"{string_of_arith(lhs)} >= {string_of_arith(rhs)}"
        case EQ(lhs, rhs):
            return f"{string_of_arith(lhs)} == {string_of_arith(rhs)}"


def string_of_stmt(s: Stmt) -> str:
    match s:
        case Assign(name, expr):
            return f"Assign({name}, {string_of_arith(expr)})"
        case Skip():
            return "Skip"
        case Block(body):
            return f"Block({string_of_stmt(body)})"
        case Seq(first, second):
            return f"{string_of_stmt(first)}; {string_of_stmt(second)}"
        case While(pred, body):
            return f"While({string_of_pred(pred)}, {string_of_stmt(body)})"
        case Print(expr):
            if isinstance(expr, (Var, Num, Add, Sub, Mul, Div)):
                return f"Print({string_of_arith(expr)})"
            return f"Print({string_of_pred(expr)})"
        case If(pred, then_branch, else_branch):
            return (
                f"If({string_of_pred(pred)}, "
                f"{string_of_stmt(then_branch)}, "
                f"{string_of_stmt(else_branch)})"
            )

File: c/syntax.c
#include "syntax.h"
#include "util.h"

#include <stdio.h>
#include <string.h>

Arith *arith_var(char *name) {
    Arith *a = xmalloc(sizeof *a);
    a->kind = ARITH_VAR;
    a->name = name;
    return a;
}

Arith *arith_num(int n) {
    Arith *a = xmalloc(sizeof *a);
    a->kind = ARITH_NUM;
    a->n = n;
    return a;
}

Arith *arith_bin(ArithKind k, Arith *lhs, Arith *rhs) {
    Arith *a = xmalloc(sizeof *a);
    a->kind = k;
    a->lhs = lhs;
    a->rhs = rhs;
    return a;
}

Pred *pred_const(PredKind k) {
    Pred *p = xmalloc(sizeof *p);
    p->kind = k;
    return p;
}

Pred *pred_not(Pred *inner) {
    Pred *p = xmalloc(sizeof *p);
    p->kind = PRED_NOT;
    p->lhs = inner;
    return p;
}

Pred *pred_log(PredKind k, Pred *lhs, Pred *rhs) {
    Pred *p = xmalloc(sizeof *p);
    p->kind = k;
    p->lhs = lhs;
    p->rhs = rhs;
    return p;
}

Pred *pred_cmp(PredKind k, Arith *a, Arith *b) {
    Pred *p = xmalloc(sizeof *p);
    p->kind = k;
    p->a = a;
    p->b = b;
    return p;
}

Stmt *stmt_assign(char *name, Arith *e) {
    Stmt *s = xmalloc(sizeof *s);
    s->kind = STMT_ASSIGN;
    s->name = name;
    s->arith = e;
    return s;
}

Stmt *stmt_skip(void) {
    Stmt *s = xmalloc(sizeof *s);
    s->kind = STMT_SKIP;
    return s;
}

Stmt *stmt_block(Stmt *body) {
    Stmt *s = xmalloc(sizeof *s);
    s->kind = STMT_BLOCK;
    s->first = body;
    return s;
}

Stmt *stmt_seq(Stmt *a, Stmt *b) {
    Stmt *s = xmalloc(sizeof *s);
    s->kind = STMT_SEQ;
    s->first = a;
    s->second = b;
    return s;
}

Stmt *stmt_while(Pred *p, Stmt *body) {
    Stmt *s = xmalloc(sizeof *s);
    s->kind = STMT_WHILE;
    s->pred = p;
    s->second = body;
    return s;
}

Stmt *stmt_print_arith(Arith *e) {
    Stmt *s = xmalloc(sizeof *s);
    s->kind = STMT_PRINT;
    s->arith = e;
    s->print_is_pred = 0;
    return s;
}

Stmt *stmt_print_pred(Pred *p) {
    Stmt *s = xmalloc(sizeof *s);
    s->kind = STMT_PRINT;
    s->pred = p;
    s->print_is_pred = 1;
    return s;
}

Stmt *stmt_if(Pred *p, Stmt *then_b, Stmt *else_b) {
    Stmt *s = xmalloc(sizeof *s);
    s->kind = STMT_IF;
    s->pred = p;
    s->first = then_b;
    s->second = else_b;
    return s;
}

char *string_of_arith(Arith *a) {
    Buf b;
    buf_init(&b);
    switch (a->kind) {
    case ARITH_VAR:
        buf_printf(&b, "Var(%s)", a->name);
        break;
    case ARITH_NUM:
        buf_printf(&b, "Num(%d)", a->n);
        break;
    case ARITH_ADD:
        buf_printf(&b, "Add(%s, %s)", string_of_arith(a->lhs), string_of_arith(a->rhs));
        break;
    case ARITH_SUB:
        buf_printf(&b, "Sub(%s, %s)", string_of_arith(a->lhs), string_of_arith(a->rhs));
        break;
    case ARITH_MUL:
        buf_printf(&b, "Mul(%s, %s)", string_of_arith(a->lhs), string_of_arith(a->rhs));
        break;
    case ARITH_DIV:
        buf_printf(&b, "Div(%s, %s)", string_of_arith(a->lhs), string_of_arith(a->rhs));
        break;
    }
    return buf_take(&b);
}

char *string_of_pred(Pred *p) {
    Buf b;
    buf_init(&b);
    switch (p->kind) {
    case PRED_TRUE:
        buf_printf(&b, "true");
        break;
    case PRED_FALSE:
        buf_printf(&b, "false");
        break;
    case PRED_NOT:
        buf_printf(&b, "not %s", string_of_pred(p->lhs));
        break;
    case PRED_AND:
        buf_printf(&b, "(%s and %s)", string_of_pred(p->lhs), string_of_pred(p->rhs));
        break;
    case PRED_OR:
        buf_printf(&b, "(%s or %s)", string_of_pred(p->lhs), string_of_pred(p->rhs));
        break;
    case PRED_LT:
        buf_printf(&b, "%s < %s", string_of_arith(p->a), string_of_arith(p->b));
        break;
    case PRED_LE:
        buf_printf(&b, "%s <= %s", string_of_arith(p->a), string_of_arith(p->b));
        break;
    case PRED_GT:
        buf_printf(&b, "%s > %s", string_of_arith(p->a), string_of_arith(p->b));
        break;
    case PRED_GE:
        buf_printf(&b, "%s >= %s", string_of_arith(p->a), string_of_arith(p->b));
        break;
    case PRED_EQ:
        buf_printf(&b, "%s == %s", string_of_arith(p->a), string_of_arith(p->b));
        break;
    }
    return buf_take(&b);
}

char *string_of_stmt(Stmt *s) {
    Buf b;
    buf_init(&b);
    switch (s->kind) {
    case STMT_ASSIGN:
        buf_printf(&b, "Assign(%s, %s)", s->name, string_of_arith(s->arith));
        break;
    case STMT_SKIP:
        buf_printf(&b, "Skip");
        break;
    case STMT_BLOCK:
        buf_printf(&b, "Block(%s)", string_of_stmt(s->first));
        break;
    case STMT_SEQ:
        buf_printf(&b, "%s; %s", string_of_stmt(s->first), string_of_stmt(s->second));
        break;
    case STMT_WHILE:
        buf_printf(&b, "While(%s, %s)", string_of_pred(s->pred), string_of_stmt(s->second));
        break;
    case STMT_PRINT:
        buf_printf(&b, "Print(%s)",
                   s->print_is_pred ? string_of_pred(s->pred) : string_of_arith(s->arith));
        break;
    case STMT_IF:
        buf_printf(&b, "If(%s, %s, %s)", string_of_pred(s->pred),
                   string_of_stmt(s->first), string_of_stmt(s->second));
        break;
    }
    return buf_take(&b);
}
